catroid/src/main/java/org/catrobat/catroid/ui/fragment/RecyclerViewListFragment.java

```diff
@@ -0,0 +1,239 @@
+
+		@Override
+		public boolean onCreateActionMode(ActionMode mode, Menu menu) {
+			MenuInflater inflater = mode.getMenuInflater();
+			inflater.inflate(R.menu.context_menu, menu);
+			return true;
+		}
+
+		@Override
+		public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
+			menu.findItem(R.id.btnEdit).setVisible(adapter.getSelectedItemCount() == 1);
+			return true;
+		}
+
+		@Override
+		public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
+
+			List<T> selectedItems = adapter.getSelectedItems();
+
+			switch (item.getItemId()) {
+				case R.id.btnEdit:
+					showRenameDialog(selectedItems.get(0).getName());
+					break;
+				case R.id.btnCopy:
+					copyItems(selectedItems);
+					break;
+				case R.id.btnDelete:
+					deleteItems(selectedItems);
+					break;
+				default:
+					return false;
+			}
+
+			return true;
+		}
+
+		@Override
+		public void onDestroyActionMode(ActionMode mode) {
+			adapter.clearSelection();
+			getActivity().invalidateOptionsMenu();
+			actionMode = null;
+		}
+	};
+
+	@Override
+	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+		setHasOptionsMenu(true);
+		view = (RecyclerView) inflater.inflate(R.layout.fragment_recycler_view, container, false);
+		return view;
+	}
+
+	@Override
+	public void onStart() {
+		super.onStart();
+		initializeAdapter();
+	}
+
+	protected abstract void initializeAdapter();
+
+	@Override
+	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
+		super.onCreateOptionsMenu(menu, inflater);
+		inflater.inflate(R.menu.options_menu, menu);
+		menu.getItem(0).setVisible(ClipboardHandler.getInstance().containsPasteableItems(getItemType()));
+	}
+
+	protected abstract Class getItemType();
+
+	protected abstract DirectoryPathInfo getCurrentDirectory();
+
+	@Override
+	public boolean onOptionsItemSelected(MenuItem item) {
+		switch (item.getItemId()) {
+			case R.id.paste:
+				pasteItems();
+				return true;
+			default:
+				return super.onOptionsItemSelected(item);
+		}
+	}
+
+	@Override
+	public void onSelectionChanged(boolean isSelectionActive) {
+		if (isSelectionActive && actionMode == null) {
+			actionMode = getActivity().startActionMode(actionModeCallback);
+			return;
+		}
+
+		if (actionMode != null) {
+			actionMode.invalidate();
+			if (!isSelectionActive) {
+				actionMode.finish();
+			}
+		}
+	}
+
+	@Override
+	public void clearSelection() {
+		actionMode.finish();
+	}
+
+	protected String getUniqueItemName(String name) {
+
+		if (isItemNameUnique(name)) {
+			return name;
+		}
+
+		int i;
+		String prefix;
+
+		try {
+			prefix = name.substring(0, name.lastIndexOf('('));
+			prefix = prefix.trim();
+			String suffix = name.substring(name.lastIndexOf('(') + 1, name.lastIndexOf(')'));
+			i = Integer.parseInt(suffix);
+			i++;
+		} catch (Exception e) {
+			prefix = name;
+			i = 1;
+		}
+
+		return getUniqueItemName(prefix + " (" + i + ")");
```

Comment:
yes, I did not see the advantage in a loop, but now that I thought about it - A loop can get a stopping condition, which would prevent us from recursing forever if something goes wrong - good point. 😼